File: leetcode/139.word-break.py
# @leet start
class Solution:
    def wordBreak(self, s: str, wordDict: list[str]) -> bool:
        wordDict.sort(key=len)

        # Is it possible to end at this exclusive index?
        dp: list[bool] = [False] * (len(s) + 1)
        dp[0] = True

        for i in range(1, len(s) + 1):
            for word in wordDict:
                if len(word) > i:
                    break

                if word == s[i - len(word) : i]:
                    dp[i] |= dp[i - len(word)]

        return dp[len(s)]


# @leet end

File: leetcode/139.word-break.js
// @leet start
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
var wordBreak = function (s, wordDict) {
  const dp = Array(s.length + 1).fill(false);
  dp[0] = true;

  for (let i = 1; i <= s.length; i++) {
    for (const word of wordDict) {
      if (word.length > i) {
        continue;
      }

      let matches = true;

      for (let j = 0; j < word.length; j++) {
        if (word[j] !== s[i - word.length + j]) {
          matches = false;
        }
      }

      dp[i] ||= matches && dp[i - word.length];
    }
  }

  return dp[s.length];
};
// @leet end
